libs/nio/src/main/java/org/elasticsearch/nio/NioSelector.java

```diff
@@ -196,7 +197,8 @@ void cleanupAndCloseChannels() {
         cleanupPendingWrites();
         channelsToClose.addAll(channelsToRegister);
         channelsToRegister.clear();
-        channelsToClose.addAll(selector.keys().stream().map(sk -> (ChannelContext<?>) sk.attachment()).collect(Collectors.toList()));
+        channelsToClose.addAll(selector.keys().stream()
```

Comment:
This started failing when a close was executed before the attachment was set on the selection key. It seems to me that this just got more likely with this change timing wise maybe, but was completely possible without it already.